### PR TITLE
feat(dropdown): add dropdown directive

### PR DESCRIPTION
**Pull request by mxth**

### CODE REVIEW

File: src/elements/dropdown/dropdown.spec.js [hunk 1]
```diff
@@ -0,0 +1,203 @@
+describe('Semantic-UI: Elements - smRating', function() {
```
Comment (RomainLanz) [inline src/elements/dropdown/dropdown.spec.js]: Wrong naming

File: src/elements/dropdown/demo/demo.js [hunk 2]
```diff
@@ -0,0 +1,286 @@
+(function() {
+    'use strict';
+
+  angular
+    .module('demo', ['semantic.ui.elements.dropdown'])
+    .controller('DropdownDemoCtrl', Ctrl);
+
+  function Ctrl() {
```
Comment (m0t0r) [inline src/elements/dropdown/demo/demo.js]: Please, use controller's name as a function name

File: src/elements/dropdown/dropdown.js [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+
+      function isType(type) {
+        return attrs.type && attrs.type.indexOf(type) > -1;
+      }
+    }
+
+    function postLink(scope, element, attrs, ctrls) {
+      var
+        dropdownCtrl = ctrls[0],
+        ngModel = ctrls[1];
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: Should we throw/warn when ng-model was not provided ?

Comment (stevesherrie) [inline src/elements/dropdown/dropdown.js]: That's the convention we have been following so far.

File: src/elements/dropdown/dropdown.js [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+    }
+  }
+
+  DropdownCtrl.$inject = ['$scope', '$element', '$attrs', '$parse', '$timeout', '$document'];
+
+  function DropdownCtrl($scope, $element, $attrs, $parse, $timeout, $document) {
+    var self = this,
+      mouseClick;
+
+    self.$scope = $scope;
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: attaching everything to _this_ is an open question here as well

File: src/elements/dropdown/dropdown.js [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+    self.$element = $element;
+    self.items = [];
+
+    self.labelElm = $element.find('sm-dropdown-label');
+
+    if ($attrs.isOpen) {
+      var getter = $parse($attrs.isOpen);
+      self.getIsOpen = function() {
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: Do not like this idea of method creation in an if condition

File: src/elements/dropdown/dropdown.js [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+        getter.assign($scope, value);
+      };
+    }
+
+    var menuElm = $element.find('sm-dropdown-menu');
+
+    $scope.$watch(function() {
+      return self.getIsOpen();
+    }, function(isOpen) {
+      $element[isOpen ? 'addClass' : 'removeClass']('active visible');
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: Not sure if this is actually readable

File: src/elements/dropdown/dropdown.js [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+
+      if (isOpen) {
+        var selectedItem = self.getSelectedItem() || self.items[0];
+
+        if (selectedItem) {
+          selectedItem.select();
+        }
+
+        $timeout(function() {
+          $document.on('click', closeDropdown);
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: why would we need to listen on document's click and not just close dropdown on select?

File: src/elements/dropdown/dropdown.js [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+            $element.triggerHandler('blur');
+            break;
+        }
+      }
+    }
+  }
+
+  DropdownCtrl.prototype.init = function(ngModel) {
+    var self = this;
+    self.ngModel = ngModel;
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: why do you attach ngModel to this? Feels like you overuse _this_

File: src/elements/dropdown/dropdown.js [hunk 9]
```diff
@@ -0,0 +1,455 @@
...
+          if (item.active) {
+            self.setLabel(item.$element.html());
+            self.labelElm.removeClass('default');
+          }
+        });
+      };
+    }
+  };
+
+  DropdownCtrl.prototype.getIsOpen = function() {
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: Not a fan doing such things.  You already have _getIsOpen_ and _setIsOpen_ methods

File: src/elements/dropdown/dropdown.js [hunk 10]
```diff
@@ -0,0 +1,455 @@
...
+    return this.isOpen;
+  };
+
+  DropdownCtrl.prototype.setIsOpen = function(value) {
+    this.isOpen = value;
+  };
+
+  DropdownCtrl.prototype.toggle = function(open) {
+    this.setIsOpen(arguments.length ? !!open : !this.getIsOpen());
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: This is an example of unreadable code. Could you improve in here ?

File: src/elements/dropdown/dropdown.js [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+    if (!this.$scope.$root.$$phase) {
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: This feels like a [really bad idea](https://github.com/angular/angular.js/wiki/Anti-Patterns)

File: src/elements/dropdown/dropdown.js [hunk 12]
```diff
@@ -0,0 +1,455 @@
...
+    });
+  };
+
+  DropdownCtrl.prototype.selectNext = function() {
+    var
+      selectedItem = this.getSelectedItem(),
+      nextIndex = this.items.indexOf(selectedItem) + 1,
+      nextItem = this.items[nextIndex];
+
+    while (nextItem && nextItem.filtered) {
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: I am not sure I understand this part

File: src/elements/dropdown/dropdown.js [hunk 13]
```diff
@@ -0,0 +1,455 @@
...
+
+  function smDropdownItem() {
+    return {
+      require: ['^smDropdown', 'smDropdownItem'],
+      controller: DropdownItemCtrl,
+      link: link
+    };
+
+    function link(scope, element, attrs, ctrls) {
+      var dropdownCtrl = ctrls[0];
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.js]: Should we throw is we do not controllers ?

File: src/elements/dropdown/dropdown.spec.js [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    leftArrow  : 37,
+    upArrow    : 38,
+    rightArrow : 39,
+    downArrow  : 40
+  };
+
+  beforeEach(module('semantic.ui.elements.dropdown'));
+
+  beforeEach(inject(function(_$rootScope_, _$compile_) {
+    $rootScope = _$rootScope_.$new();
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.spec.js]: $rootScope.$new() creates a new scope which is not going to be _$rootScope_

File: src/elements/dropdown/dropdown.spec.js [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    element.triggerHandler('focus');
+    expectOpen();
+  });
+
+  it('select item', function() {
+    dropdownClick();
+    getItem(0).click();
+    expect($rootScope.selected).toEqual('0');
+    expect(getStateActive()).toEqual([true, false, false]);
+    expect(getLabel().html()).toEqual(getItem(0).html());
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.spec.js]: Example of unreadable expectation. Could improve it, please ?

File: src/elements/dropdown/dropdown.spec.js [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    element.click();
+  }
+
+  function triggerKeyDown(keyCode) {
+    var e = $.Event('keydown');
+    e.keyCode = keyCode;
+    element.trigger(e);
+  }
+
+  function expectOpen() {
```
Comment (m0t0r) [inline src/elements/dropdown/dropdown.spec.js]: Not supporting this idea to wrap expectations to a method

**Comment by m0t0r** [comment]
@mxth I felt pretty uncomfortable reading some parts of the code. Can you review it and possibly improve readability, please ? Please, check my other comments

**Pull request closed without merging**